crypto/bio/bio_lib.c

```diff
@@ -848,29 +857,20 @@ int BIO_connect_retry(BIO *bio, int timeout)
         return 0;
     }
 
+    if (polling_milliseconds < 0)
+        polling_milliseconds = 100;
     if (!blocking)
         BIO_set_nbio(bio, 1);
 
- retry: /* it does not help here to set SSL_MODE_AUTO_RETRY */
-    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+ retry: /* it does not always help to set SSL_MODE_AUTO_RETRY */
+    clear_sys_error();
+    rv = BIO_do_connect(bio); /* This indirectly calls ERR_clear_error(); */
 
     if (rv <= 0) {
-        if (get_last_sys_error() == ETIMEDOUT) {
-            /*
-             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
-             */
-            ERR_clear_error();
-            (void)BIO_reset(bio);
-            /*
-             * unless using BIO_reset(), blocking next connect() may crash and
-             * non-blocking next BIO_do_handshake() will fail
-             */
-            goto retry;
-        } else if (BIO_should_retry(bio)) {
-            /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
-            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+        if (BIO_should_retry(bio)
+                || get_last_sys_error() == EAGAIN /* Temporary gai error */) {
```

Comment:
What is a "gai" error?
and if we end up here with EAGAIN and !BIO_should_retry,
that looks like a bug, so why not simply fix the bug?

Reply:
By "gai" error I meant the following `getaddrinfo()` error: Temporary failure in name resolution.
I've improved the commenting and the handling of this case in my most recent push.
This is all I can do today; anything else needs to wait until tomorrow.